### PR TITLE
Actor Registry

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: quickwit/quickwit-actors/src/actor.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ use tracing::{debug, error, info_span, Span};
 
 use crate::actor_state::{ActorState, AtomicState};
 use crate::progress::{Progress, ProtectedZoneGuard};
+use crate::register::ActorRegister;
```
Comment (guilload) [inline quickwit/quickwit-actors/src/actor.rs:36]:
```suggestion
use crate::registry::ActorRegistry;
```

Comment (fulmicoton) [reply quickwit/quickwit-actors/src/actor.rs]: renamed

**Review by boraarslan — COMMENTED** [review]
Looks great; I have two questions though.
1. Why are we using `std::sync::RwLock` instead of tokio's `RwLock`? Is there any particular reason?
2. Unwrapping a poisoned lock would result in a crash. Is this what we want, or do we have a recovery logic?

**Comment by fulmicoton** [comment]
> Looks great; I have two questions though.
> 
> 1. Why are we using `std::sync::RwLock` instead of tokio's `RwLock`? Is there any particular reason?
> 2. Unwrapping a poisoned lock would result in a crash. Is this what we want, or do we have a recovery logic?

Two great questions...

I'll start by 2.
Absolutely... In tantivy and quickwit, we try to make that kind of lock usage as trivial as possible.
This means: ideally, keeping your lock guard in its simple scope (not returning it, etc).
Ideally only execute trivial non-panicking code while the lock is held. Often this means we take the lock, clone the value and unlock. This is not great from a performance standpoint but this is much easier to read proof.
It is tempting to just use a magical implementation that cannot poison but the problem is not really solved is it?
If your tokio task panic while you where writing your object you might have left your object in an inconsistent state!
Panicking on the next lock acquisition is actually better than letting readers seeing (and possibly propagating/persisting etc) that inconsistent state.

1. The async Mutex is required if you have the lock accross an async.
https://tokio.rs/tokio/tutorial/shared-state#on-using-stdsyncmutex

Outside of this, if something on which there are no contention like it is the case here, using sync::Mutex is fine, and actually more efficient.
By using a std::sync::Mutex I'm expressing that this is supposed to be a trivial case, and the reader should assume:
- a small scope
- only simple computation done (no IO for instance)

File: quickwit/quickwit-actors/src/lib.rs [hunk 2]
```diff
@@ -91,3 +93,13 @@ impl<E: fmt::Debug + ServiceError> ServiceError for AskError<E> {
...
+
+struct TestActor;
+
+impl Actor for TestActor {
+    type ObservableState = usize;
+
+    fn observable_state(&self) -> Self::ObservableState {
+        0
+    }
+}
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-actors/src/lib.rs]: agreed

File: quickwit/quickwit-actors/src/registry.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            }));
+    }
+
+    // As far as I can tell this is a false alarm from clippy.
+    #[allow(clippy::await_holding_lock)]
+    pub async fn observe(&self, timeout: Duration) -> Vec<ActorObservation> {
+        self.gc();
+        let rlock = self.actors.read().unwrap();
+        let mut obs_futures = Vec::new();
+        for registry_for_type in rlock.values() {
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-actors/src/registry.rs]: ok I switched to this and removed the clippy attribute

File: quickwit/quickwit-actors/Cargo.toml [hunk 4]
```diff
@@ -18,9 +18,10 @@ futures = { workspace = true }
 thiserror = { workspace = true }
 tokio = { workspace = true }
 tracing = { workspace = true }
-
 quickwit-common = { workspace = true }
 quickwit-proto = { workspace = true }
+serde = { workspace = true }
+serde_json = { workspace = true }
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-actors/Cargo.toml]: ok

**Review by boraarslan — APPROVED** [review]

**Pull request merged**